src/databricks/labs/ucx/source_code/notebook_linter.py

```diff
@@ -11,15 +11,31 @@ class NotebookLinter:
     to the code cells according to the language of the cell.
     """
 
-    def __init__(self, languages: Languages):
-        self._languages = languages
+    def __init__(self, langs: Languages, notebook: Notebook):
+        self._languages: Languages = langs
+        self._notebook: Notebook = notebook
+        self._cell_offsets: list[int] = []
 
-    def lint(self, notebook: Notebook) -> Iterable[Advice]:
-        for cell in notebook.cells:
+    @classmethod
+    def from_source(cls, langs: Languages, source: str, default_language: Language) -> 'NotebookLinter':
+        notebook = Notebook.parse("", source, default_language)
+        assert notebook is not None
+        return cls(langs, notebook)
+
+    def lint(self) -> Iterable[Advice]:
+        for cell in self._notebook.cells:
             if not self._languages.is_supported(cell.language.language):
                 continue
             linter = self._languages.linter(cell.language.language)
-            yield from linter.lint(cell.original_code)
+            for advice in linter.lint(cell.original_code):
+                yield advice
+                self._cell_offsets.append(cell.original_offset)
+
+    def adjust_advices(self, advices: list[Advice]) -> Iterable[Advice]:
+        for advice, offset in zip(advices, self._cell_offsets):
+            advice = advice.replace(start_line=advice.start_line + offset, end_line=advice.end_line + offset)
```

Comment:
Do we even need this method? Why can't we replace line offsets before yielding?

Reply:
You originally asked for reporting at cell offsets, then a layer to do the translation. If we do not need the original cell offsets, then I can remove this logic and just do it when the advice is first raised.

Reply:
Changed to only yield adjusted offsets